engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java

```diff
@@ -880,6 +898,19 @@ private void execute() throws YarnException, IOException
         if (0 != exitStatus) {
           if (allocatedContainer != null) {
             numFailedContainers.incrementAndGet();
+            if (exitStatus != 1) {
+              // If container failure due to framework
+              String hostname = allocatedContainer.container.getNodeId().getHost();
+              int failedTimes = failedContainersMap.putIfAbsent(hostname, 1);
+              if (failedTimes > 1) {
+                failedContainersMap.replace(hostname, ++failedTimes);
```

Comment:
Generally speaking not thread safe if this function will be invoked from two different threads with containers pointing to the same hostname.